src/user/UserCredentials.tsx

```diff
@@ -150,6 +158,163 @@ export const UserCredentials = ({ user }: UserCredentialsProps) => {
+    arr.splice(toIndex, 0, temp[0]);
+    return arr;
+  };
+
+  const move = (itemOrder: string[]) => {
+    if (!bodyRef.current) return;
+    const ulNode = bodyRef.current;
+    const nodes = Array.from(ulNode.children);
+    if (nodes.map((node) => node.id).every((id, i) => id === itemOrder[i])) {
+      return;
+    }
+    while (ulNode.firstChild) {
+      ulNode.removeChild(ulNode.lastChild!);
+    }
+    itemOrder.forEach((id) => {
+      ulNode.appendChild(nodes.find((n) => n.id === id)!);
+    });
+  };
+
+  const onDragCancel = () => {
+    Array.from(bodyRef.current?.children || []).forEach((el) => {
+      el.classList.remove(styles.modifiers.ghostRow);
+      el.setAttribute("aria-pressed", "false");
+    });
+    setState({
+      ...state,
+      draggedItemId: "",
+      draggingToItemIndex: -1,
+      dragging: false,
+    });
+  };
+
+  const onDragLeave = (evt: React.DragEvent) => {
+    if (!isValidDrop(evt)) {
+      move(itemOrder);
+      setState({ ...state, draggingToItemIndex: -1 });
+    }
+  };
+
+  const isValidDrop = (evt: React.DragEvent) => {
+    const ulRect = bodyRef.current!.getBoundingClientRect();
+    return (
+      evt.clientX > ulRect.x &&
+      evt.clientX < ulRect.x + ulRect.width &&
+      evt.clientY > ulRect.y &&
+      evt.clientY < ulRect.y + ulRect.height
+    );
+  };
+
+  const onDrop = (evt: React.DragEvent) => {
+    if (isValidDrop(evt)) {
+      onDragFinish(state.draggedItemId, state.tempItemOrder);
+    } else {
+      onDragCancel();
+    }
+  };
+
+  const onDragOver = (evt: React.DragEvent) => {
+    evt.preventDefault();
+    const td = evt.target as HTMLTableCellElement;
+    const curListItem = td.closest("tr");
+    if (
+      !curListItem ||
+      (bodyRef.current && !bodyRef.current.contains(curListItem)) ||
+      curListItem.id === state.draggedItemId
+    ) {
+      return null;
+    } else {
+      const dragId = curListItem.id;
+      const draggingToItemIndex = Array.from(
+        bodyRef.current?.children || []
+      ).findIndex((item) => item.id === dragId);
+      if (draggingToItemIndex !== state.draggingToItemIndex) {
+        const tempItemOrder = moveItem(
+          itemOrder,
+          state.draggedItemId,
+          draggingToItemIndex
+        );
+        move(tempItemOrder);
+        setState({
+          ...state,
+          draggingToItemIndex,
+          tempItemOrder,
+        });
+      }
+    }
+  };
+
+  const onDragEnd = (evt: React.DragEvent) => {
+    const tr = evt.target as HTMLTableRowElement;
+    tr.classList.remove(styles.modifiers.ghostRow);
+    tr.setAttribute("aria-pressed", "false");
+    setState({
+      ...state,
+      draggedItemId: "",
+      draggingToItemIndex: -1,
+      dragging: false,
+    });
+  };
+
+  const onDragFinish = async (_dragged: string, newOrder: string[]) => {
+    const dragged = _dragged.split(",")[0];
+    const keys = groupedUserCredentials.map((e) => e.value.map((c) => c.id));
```

Comment:
You can make this code a bit more readable by dropping the single letter variables and using destructuring instead.
```suggestion
    const keys = groupedUserCredentials.map(({ value }) =>
      value.map(({ id }) => id)
    );
```